src/basins/basins_types.jl

```diff
@@ -0,0 +1,155 @@
+
+A subtype of [`BasinsOfAttraction`](@ref) whose `basins` of attraction are represented by an array, 
+or more specifically a subtype of `AbstractArray` whose values have type `ID` and has `D` 
+number of dimensions. The `attractors` take the form of a dictionary mapping attractor labels
+to `StateSpaceSet`'s with the points of each set being of length `D`. The domain here is `grid`
+and has type `Grid`, each grid "cell" has a unique corresponding entry in the basins. 
+
+The first constructor simply converts the tuple representation of the grid into type `Grid`, 
+and then calls the second constructor. This `grid` tuple is a tuple of ranges defining the grid 
+of initial conditions that partition the state space. The `grid` has to be of the same dimensionality
+as the points representing the attractors and of the `basins`. 
+"""
+struct ArrayBasinsOfAttraction{ID, D, B <: AbstractArray{ID,D}, T, V, G <: Grid, AK, S <: StateSpaceSet{D, T, V}} <: BasinsOfAttraction{ID}
+    basins::B 
+    attractors::Dict{AK, S}
+    grid::G
+end
+# The definition of the first constructor can be found in `basins/basins_utilities.jl`.
+
+"""
+    SampledBasinsOfAttraction{ID, D} <: BasinsOfAttraction{ID}
+    SampledBasinsOfAttraction(basins, attractors, sampled_points::Vector{U}) where {U <: AbstractVector}
+    SampledBasinsOfAttraction(basins, attractors, sampled_points::S) where {S <: AbstractStateSpaceSet}
+
+A subtype of [`BasinsOfAttraction`](@ref) whose `basins` of attraction are represented by a vector
+whose values are of type `ID`. The `attractors` take the form of a dictionary mapping attractor 
+labels to `StateSpaceSet`'s with the points of each set being of length `D`. The domain of this 
+basin type is `sampled_points` which is also a `StateSpaceSet` with the same dimensionality `D`, 
+element type, and vector type as those used to represent the attractors. Each sampled point has a 
+unique corresponding entry in the `basins` vector. The length of `basins` and `sampled_points` must
+be equal.
+
+The first constructor converts a vector of vectors into a `StateSpaceSet`, then calls the second constructor.
+That is, assuming each has the same dimensionality `D`, and the same element type as the `StateSpaceSet`'s 
+representing the attractors.
+"""
+struct SampledBasinsOfAttraction{ID, D, T, V <: AbstractVector, AK, S <: StateSpaceSet{D, T, V}} <: BasinsOfAttraction{ID}
+    points_ids::Vector{ID}
+    attractors::Dict{AK, S}
+    sampled_points::S
+end
+# The definition of the first constructor can be found in `basins/basins_utilities.jl`.
+
+#########################################################################################
+# Basins of Attraction Convenience functions
+######################################################################################### 
+Base.iterate(BoA::BasinsOfAttraction, state=1) = state == 1 ? (extract_basins(BoA), 2) : state == 2 ? (extract_attractors(BoA), 3) : nothing
```

Comment:
Yes, this is great. this way, the `basins_of_attraction` function can return `BoA` types and not be breaking.